src/discord_bot/bot.cpp

```diff
@@ -3,23 +3,176 @@
 
 #include "eelbot_framework/discord_bot/bot.hpp"
 
+#include "eelbot_framework/compression.hpp"
+#include "eelbot_framework/discord_bot/endpoints.hpp"
 #include "eelbot_framework/discord_bot/rest_api.hpp"
 #include "eelbot_framework/json.hpp"
+#include "eelbot_framework/os.hpp"
+#include "eelbot_framework/websocket.hpp"
 #include "fmt/core.h"
 
 #include <stdexcept>
+#include <thread>
 namespace eelbot_framework {
 
 namespace discord_bot {
 
-http_request_settings bot::get_http_request_settings() {
+http_request_settings bot::get_http_request_settings() noexcept {
 	http_request_settings request_settings;
 	request_settings.tls_ver      = this->tls_ver;
 	request_settings.ca_directory = this->ca_directory;
 	request_settings.proxy        = this->http_proxy;
 	return request_settings;
 }
 
+void bot::send_heartbeat() {
+	payload pl;
+	pl.op = opcode::heartbeat;
+	pl.d  = this->sequence_num;
+	this->ws_client->send(to_json_str(pl));
+}
+
+void bot::send_identify() {
+	// TODO: Make shard info configurable down the line.
+	shard_info sh_info;
+	sh_info.shard_id   = 0;
+	sh_info.num_shards = 1;
+
+	identify_connection_properties id_conn_props;
+	id_conn_props.os      = get_platform_name();
+	id_conn_props.browser = "Eelbot Framework";
+	id_conn_props.device  = "Eelbot Framework";
+
+	identify id;
+	id.token      = this->token;
+	id.intents    = 513; // TODO: Determine this dynamically.
+	id.shard      = sh_info;
+	id.properties = id_conn_props;
+	id.compress   = false;
+
+	this->logger->log(log::level::DEBUG, "Sending identify.");
+	payload pl;
+	pl.op = opcode::identify;
+	pl.d  = id;
+	this->ws_client->send(to_json_str(pl));
+}
+
+void bot::send_resume() {
+	resume rs;
+	rs.token = this->token;
+	rs.seq   = this->sequence_num;
+
+	{
```

Comment:
Whats the purpose behind something like this scoped block?

Reply:
This section uses a mutex to avoid a race condition. The best practice approach to locking and unlocking mutexes is not to call mutex.lock() and mutex.unlock() directly, but rather do it through a scoped lock. A scoped lock takes in a mutex during construction and automatically locks it. Upon it's destruction, it calls unlock on the underlying mutex.

The reason it is preferred to use this is so if exceptions are thrown before calling mutex.unlock(), this object is always guaranteed to be destructed when going out of scope and thus the mutex will never get deadlocked in a locked state.

For more information: https://en.cppreference.com/w/cpp/thread/scoped_lock